File: samples/fern.js
const { create_turtle, penup, pendown, penwidth, clear, stop, pencolour, forward, back, font, filltext, stroketext, right, left, print } = require('@picoturtle/picoturtle-nodejs-client');

let main = async () => {
    await create_turtle();
    /* Your code goes here */

    async function fern(size, sign) {
        if (size < 1) {
            return;
        }
        await forward(size);

        await right(70 * sign);
        await fern(size/2, sign*-1);
        await left(70 * sign);

        await forward(size);

        await left(70 * sign);
        await fern(size/2, sign);
        await right(70 * sign);

        await right(7 * sign);
        await fern(size - 2, sign);
        await left(7 * sign);

        await forward(-size);
        await forward(-size);
    }

    await penup();
    await back(200);

    await right(90);
    await back(20);
    await pencolour(200, 200, 0);
    await font('14pt Arial');
    await filltext('Fern');
    await forward(20);
    await left(90);
    await forward(50);

    await pencolour(0, 128, 0);
    await pendown();
    await fern(25, 1);

    /* Your code ends here */

    /* Always stop the turtle */
    await stop();
};
main();

File: lua/samples/fern.lua
-- fern.lua
-- Author: Abhishek Mishra
-- Date: 15/01/2022
--
-- Draws a fern
--

-- fern: function which
-- calls itself recursively
-- till the size is less than 1
-- to draw the fern.

local picoturtle = require "picoturtle"

-- if called from another script, t will be set to a turtle object
-- if t is not set to a turtle (i.e. it is nil) then create a new one.
local t = t or picoturtle.new()

-- reset turtle in case we are reusing a passed turtle object
t:reset()

function fern(size, sign)
    if (size < 1) then
        return
    end
    
    t:forward(size)

    t:right(70 * sign)
    fern(size/2, sign*-1)
    t:left(70 * sign)

    t:forward(size)

    t:left(70 * sign)
    fern(size/2, sign)
    t:right(70 * sign)

    t:right(7 * sign)
    fern(size - 2, sign)
    t:left(7 * sign)

    t:forward(-size)
    t:forward(-size)
end

t:penup()
t:back(200)

t:right(90)
t:back(20)
t:pencolour(200, 200, 0)
t:font('Calibri', 50)
t:filltext('Fern')
t:forward(20)
t:left(90)
t:forward(50)

t:pencolour(128, 128, 0)
t:pendown()
fern(25, 1)
